### PR TITLE
Copy over Mainstream browse pages into Topics

### PR DESCRIPTION
**Pull request by AgaDufrat**
## Context
Link to Trello card: https://trello.com/c/KGCUHfhi/876-copy-over-mainstream-browse-pages-into-specialist-topics

# Changes proposed in this pull request

Add a rake task to copy over Mainstream Browse Pages (MBP) into Topics.
The task is intended to run once.

Tested locally with a copy of production data. Completed in ~ 40 minutes

# Assumptions / Scope

- At this point we don't copy over MBP's attributes which would require Topic content schema changes such as equivalent of `top_level_browse_pages`. We decided that those changes can be later once we identify what we need

- Only published MBP will be cloned
The product team decided that we will not copy over archived ones. I decided not to copy draft MBP as there is not much value in keeping those in sync. I think a MBP should be copied into a Topic when the MBP is published.

- We assume that there are no duplicates, we naively copy all MBPs.

# How it works

- First, **create basic new Topics**
    - the new Topic is at this point associated to the MBP parent because the corresponding Topic parent may not exist yet
    - the slug is appended with `-mainstream-copy` as there is a unique slug constaing. This will like need to be removed once we finish the work and the original MBP is unpublished/deleted
    - :construction: **this will change** - new Topics have a flag that it used to be a MBP 

- **Tag documents that are tagged to the MBP to the new Topic**
    This is the slowest operation. As far as I know there is no way to simply add a new link to topic or tag all documents to the Topic in one call. Therefore for each document (there is ~3.5k documents tagged to MBPs) it:
    - calls  the Publishing API to get document's current links
    - adds the link to the new Topic
    - sends the new links to the Publishing API
    
- **Copy curated lists and list items**
    when there are no lists `"groups": []`

After those operations:

- **Update parent and child associations between the new Topic**
    When the Tag/Topic is published, it ustilises the `DependentTagPublishWorker` in the `PublishingAPINotifier` to ensure the "reverse linking"

- **Send everything to Publishing API**
    If the MBP has curated lists, this can be making a lot of calls to Publishing API (`put_content` and `publish` with empty groups and with populated groups). Curated lists/groups are published separately to the Topics itself. The Topic needs to be published before the ListPublisher is called as it fetches the tagged_documents utilising `Services.publishing_api.get_linked_items`
If we don't publish the Topic first we'll get 404 No item with content_id of the new Topic.

# Guidance to review
    
Commit by commit.

> I'm aware that this is not the prettiest of codes and could do with refactoring to confirm to best object oriented design practices. The private methods should be refactored to Services, which can be used to clone MPB into a Topic when a MBP is published. It'll be quicker (less tests to change) to do it after the initial functionality is reviewed.

To test locally:
1. Replicate production data for publishing-api, content-store and collections-publisher
2. Call the rake task or service in rails console in docker

# Things to check and review questions  
- Are there any other tags/keys in "links" in we need to carry over? E.g. “related_content”
- Is there a way of reducing the number of calls Publishing API? Would it be worth the effort.

**Notes:**
- I suspect we will also need to reference the cloned topic on the original MBP. This will clarify while we are working on syncing the document tagging ("_When a document is tagged to MBP, also tag it to the clone Topic_")
- If we decide to first start syncing any *new* MBPs, and then use this to backfill the data we will need to add a line to skip if the cloned Topic exists
- TODO: Refactor
  - make private methods separate services
  - add a namespace so it's easier to delete

# Depends on
- Hiding the Browse topic from publishing apps: Whitehall, Publisher, Collections Publisher, Content tagger being completed 
- Hide the new Browse topics from the frontend apps being completed 

---
⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

### CODE REVIEW

**Review by hannako — COMMENTED** [review]
This is looking really fantastic Aga, massive props for getting this written and running. Such a lot of stuff to think about. The PR overview is very helpful - it made reviewing so much easier. 🥇 💪 

Hopefully the comments above make sense. Happy to chat through any of it if that's helpful.

With regard to your questions: 

> Are there any other tags/keys in "links" in we need to carry over? E.g. “related_content”

 I think that will depend on how we will be rendering topic pages in the new world, ie the results of the AB test. I can't find a MSB with a `related_content` key, do you mean `related_topics`? If so it looks like they are used for the Detailed Guidance part of the page on [second level browse page](https://github.com/alphagov/collections/blob/1051472eabc97fbc6068eef846a1fcdfde5b8c1c/app/views/second_level_browse_page/_links.html.erb#L48)s. We can check against the prototype if this content is represented in the new templates. If not, probably leave it out of the import for now. 

> Is there a way of reducing the number of calls to Publishing API? Would it be worth the effort?

I think it _would_ be worth the effort, when it comes to running this in production. Perhaps it will be easier to spot areas for consolidation once you've had a chance to refactor.

**Comment by chao-xian** [comment]
Looks pretty good barring Jess's comments to address. Definitely worth thinking about how we reduce calls to publishing_api as the production one is prone to timeouts from what I hear.

File: app/presenters/topic_presenter.rb [hunk 1]
```diff
@@ -12,4 +12,10 @@ def links
       "primary_publishing_organisation" => [GDS_CONTENT_ID],
     )
   end
+
+  def details
+    super.merge(
+      "mainstream_browse_type" => @tag.mainstream_browse_type?,
```
[the remark this answers is not in the document]
Comment (AgaDufrat) [reply app/presenters/topic_presenter.rb]: 👍  Changed to use that the attribute reader. 
We decided not to use `mainstream_browse_type` (at least at this stage).

File: spec/services/copy_mainstream_browse_pages_to_topics_spec.rb [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+          "title" => "Carers",
+          "base_path": "/topic/carers-mainstream-copy",
+          "document_type" => "topic",
+          "schema_name": "topic",
+        }),
+      )
+    end
+  end
+
+  def stub_publishing_api_has_no_linked_items(content_id)
```
[the remark this answers is not in the document]
Comment (AgaDufrat) [reply spec/services/copy_mainstream_browse_pages_to_topics_spec.rb]: :+1: Changed to use that method

File: spec/features/curating_topic_contents_spec.rb [hunk 3]
```diff
@@ -114,6 +114,7 @@
               },
             ],
             "internal_name" => "Oil and Gas / Offshore",
+            "mainstream_browse_type": false,
```
[the remark this answers is not in the document]
Comment (AgaDufrat) [reply spec/features/curating_topic_contents_spec.rb]: We decided not to use `mainstream_browse_type` (at least at this stage) and the content schema was changed to remove this attribute.

File: app/services/copy_mainstream_browse_pages_to_topics.rb [hunk 4]
```diff
@@ -39,6 +40,16 @@ def create_basic_topic(mainstream_browse_page)
     topic if topic.save!
   end
 
+  def update_assossiations_to_topics(topic)
```
[the remark this answers is not in the document]
Comment (AgaDufrat) [reply app/services/copy_mainstream_browse_pages_to_topics.rb]: Agree that method name is not very descriptive. I called it before I knew exactly that it needs to do and never came back to review. 

I guess I wanted to highlight that as a result, the links of both parent and child topics are updated. From Publishing API point of view it tags new topic child to the new topic parent AND new topic parent to the new topic parent.
I’d probably call it `update_parent_child_associations_to_new_topics` which is even longer :joy:
This is documented in the commit message and the test, so I’ll go with `update_parent_to_new_topic` as I’d prefer to save the word “tag” for describing links to the documents.

File: lib/tasks/publishing_api.rake [hunk 5]
```diff
@@ -84,4 +84,11 @@ namespace :publishing_api do
     )
     puts "Links patched for root page..."
   end
+
+  desc "Copy all published Mainstream browse pages to Topics"
+  task copy_mainstream_browse_pages_to_topics: :environment do
```
[the remark this answers is not in the document]
Comment (AgaDufrat) [reply lib/tasks/publishing_api.rake]: Implemented in in [Ensure copy_mainstream_browse_pages_to_topics task can't run in prod](https://github.com/alphagov/collections-publisher/pull/1571/commits/f6f11682128fa585085e2714b67fdff63edaa7ef)

File: lib/tasks/browse_topics.rake [hunk 6]
```diff
@@ -4,6 +4,8 @@ require_relative "../special_route_publisher"
 namespace :browse_topics do
   desc "Copy all published Mainstream browse pages to Topics"
   task copy_mainstream_browse_pages_to_topics: :environment do
+    raise "This rake task is not intended to be run in production" if Rails.env.production?
```
Comment (hannako) [inline lib/tasks/browse_topics.rake:7]: Nice

**Review by hannako — APPROVED** [review]
Thanks for making those amends. This is looking really good.

**Pull request merged**